### PR TITLE
Fix broken explicit ask pattern for Java (closes #24587)

### LINKED ISSUE #24587: PatternCS.ask overload taking a function from ActorRef not selected by the compiler
``` java
import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.pattern.PatternsCS;
import akka.util.Timeout;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;

public class JavaMain {

    public static class ActorA extends AbstractActor {

        public ActorA(final ActorRef actorB) {
            CompletionStage<Object> response = PatternsCS.ask(
                    actorB,
                    replyTo -> new ActorB.Request("hello", replyTo),
                    Timeout.apply(1, TimeUnit.SECONDS));
            response.whenComplete((o, e) -> {
                if (o != null) System.out.println(o);
                if (e != null) System.out.println(e);
            });
        }

        @Override
        public Receive createReceive() {
            return emptyBehavior();
        }
    }

    public static class ActorB extends AbstractActor {

        public static class Request {

            public Request(final String text, ActorRef replyTo) {
                this.text = text;
                this.replyTo = replyTo;
            }

            public final String text;
            public final ActorRef replyTo;
        }

        public Receive createReceive() {
            return receiveBuilder()
                    .match(Request.class, request -> request.replyTo.tell(request.text, getSelf()))
                    .build();
        }
    }

    public static void main(String[] args) {
        final ActorSystem system = ActorSystem.create();
        final ActorRef actorB = system.actorOf(Props.create(ActorB.class), "actorB");
        final ActorRef actorA = system.actorOf(Props.create(ActorA.class, actorB), "actorA");
    }
}
```

Compiling with Java 8 and Java 9 and running results in:

```
akka.pattern.AskTimeoutException: Ask timed out on [Actor[akka://default/user/actorB#1415804947]] after [1000 ms]. Sender[null] sent message of type "akka.pattern.PatternsCS$$$Lambda$482/1291329720".
```

So obviousliy instead of selecting the "special" ask overload the "normal" one is selected and the function is not applied but instead sent as message.

### PR DESCRIPTION
**Pull request by hseeberger**

### CODE REVIEW

**Comment by hseeberger** [comment]
See #24587.

**Comment by hseeberger** [comment]
WTF! Jenkins failed with "0 failed"?! ("Test FAILed. 7310 tests run, 467 skipped, 0 failed." )

**Comment by raboof** [comment]
It's not the tests that failed, but the MiMa check:
```
[info] akka-actor: found 2 potential binary incompatibilities while checking against com.typesafe.akka:akka-actor_2.11:2.5.3  (filtered 10)
[error]  * method ask(akka.actor.ActorRef,akka.japi.Function,akka.util.Timeout)java.util.concurrent.CompletionStage in object akka.pattern.PatternsCS in current version does not have a correspondent with same parameter signature among (akka.actor.ActorRef,java.lang.Object,akka.util.Timeout)java.util.concurrent.CompletionStage, (akka.actor.ActorSelection,java.lang.Object,Long)java.util.concurrent.CompletionStage, (akka.actor.ActorSelection,java.lang.Object,akka.util.Timeout)java.util.concurrent.CompletionStage, (akka.actor.ActorRef,java.lang.Object,Long)java.util.concurrent.CompletionStage, (akka.actor.ActorSelection,akka.japi.Function,Long)java.util.concurrent.CompletionStage
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.pattern.PatternsCS.ask")
[error]  * method ask(akka.actor.ActorRef,akka.japi.Function,Long)java.util.concurrent.CompletionStage in object akka.pattern.PatternsCS in current version does not have a correspondent with same parameter signature among (akka.actor.ActorRef,java.lang.Object,akka.util.Timeout)java.util.concurrent.CompletionStage, (akka.actor.ActorSelection,java.lang.Object,Long)java.util.concurrent.CompletionStage, (akka.actor.ActorSelection,java.lang.Object,akka.util.Timeout)java.util.concurrent.CompletionStage, (akka.actor.ActorRef,java.lang.Object,Long)java.util.concurrent.CompletionStage, (akka.actor.ActorSelection,akka.japi.Function,Long)java.util.concurrent.CompletionStage
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.pattern.PatternsCS.ask")
```

**Comment by raboof** [comment]
(semi-related: really looking forward to seeing how https://github.com/akka/akka-http/pull/865 turns out, that could make this kind of feedback much nicer/easier to find)

**Comment by hseeberger** [comment]
So what should we do here? The overload was broken. But we can't remove it, due to compatibility? Even though it did not work and there were no tests?

File: akka-actor/src/main/scala/akka/pattern/Patterns.scala [hunk 1]
```diff
@@ -328,9 +336,13 @@ object PatternsCS {
...
    *   timeout);
    * }}}
+   *
+   * @param actor          the actor to be asked
+   * @param messageFactory function taking an actor ref and returning the message to be sent
+   * @param timeout        the timeout for the response before failing the returned completion stage
    */
-  def ask(actor: ActorRef, messageFactory: japi.Function[ActorRef, Any], timeout: Timeout): CompletionStage[AnyRef] =
-    scalaAsk(actor, messageFactory.apply _)(timeout).toJava.asInstanceOf[CompletionStage[AnyRef]]
+  def explicitAsk(actor: ActorRef, messageFactory: japi.function.Function[ActorRef, Any], timeout: Timeout): CompletionStage[AnyRef] =
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/pattern/Patterns.scala]: Not sure about the name, how is an ask with a factory more "explicit" than an ask with just a message? What about `askWithActorRef`?

**Review by johanandren — COMMENTED** [review]
I think we can filter the bincomp warning as the function was never callable from any Java code.

**Comment by hseeberger** [comment]
@johanandren "filter the bincomp warning" sounds good. How can we do that?

File: akka-actor/src/main/scala/akka/pattern/Patterns.scala (hunk 1 above)
Comment (hseeberger) [reply akka-actor/src/main/scala/akka/pattern/Patterns.scala]: I agree. Actually I simply took the name of the trait which provides this feature. Should we rename that, too?

**Comment by johanandren** [comment]
The statement to do it is in the log output Raboof referred to, there is a per module `mima-filters` directory where you create a file named `2.5.10.backwards.excludes` and add those statements

**Comment by hseeberger** [comment]
Thanks. Learnt something new :-)
Changes underway.

File: akka-actor/src/main/mima-filters/2.5.10.backwards.excludes [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# #24609 Fix broken explicit ask pattern for Java
+ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.pattern.PatternsCS.ask")
+ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.pattern.PatternsCS.ask")
```
Comment (jrudolph) [inline akka-actor/src/main/mima-filters/2.5.10.backwards.excludes]: One should be enough ;)

File: akka-actor/src/main/scala/akka/pattern/Patterns.scala (hunk 1 above)
Comment (johanandren) [reply akka-actor/src/main/scala/akka/pattern/Patterns.scala]: I don't find the trait name that important as that is internal and pretty simple/isolated.

File: akka-actor/src/main/mima-filters/2.5.10.backwards.excludes (hunk 2 above)
Comment (hseeberger) [reply akka-actor/src/main/mima-filters/2.5.10.backwards.excludes]: Right ;-)
Updating the PR ...

**Comment by patriknw** [comment]
@hseeberger bump

**Comment by johanandren** [comment]
I'll pick this up and push it the last bit.

**Comment by johanandren** [comment]
Continued in #24882

**Pull request closed without merging**